rip out formula parsing and replace it with something sane

kleinschmidt opened the pull request:
This PR replaces the spaghetti code that handled the expression re-writes that transform formula surface syntax into a bunch main effects and interactions.  It works by specifying re-write rules (represented by subtypes of `FormulaRewrite`) for associative rule, distributive rule, "star expansion" (`a*b` to `a+b+a&b`), and subtraction (replacing `x-1` with `x + -1`).  These rules are implemented via methods on two functions: `applies(ex::Expr, child_idx::Int, ::Type{T<:FormulaRewrite})` and `rewrite!(ex::Expr, child_idx::Int, ::Type{T<:FormulaRewrite})`.  The first checks whether the child of `ex` at `child_idx` should be re-written under rule `T`, and the second modifies `ex` in place and returns the index of the next child to check.  For example, the associative rule checks whether `ex` and `ex.args[child_idx]` are both calls to the same associative operator, and splices the args of the child into `ex` in place of the child.

The goal here is to make the code easier to read, maintain, and extend.  Be expressing all the operations on the formula DSL in the same terms it'll be easier, I hope, to expand it or for package authors to implement custom extensions (although that may require using custom macros with the current implementation).  In the future I plan to change how the terms themselves are represented to make things even more composable and extensible, but I think the current PR should be considered regardless of that.

The only thing that's currently broken is the `drop_term!` tests: they're `Terms` are equivalent but the actual `Formulae` are not because I wasn't sure what the best way to modify the whole expression of the formula is (which I now store on the `Formula` struct, to hold onto both the _original_ (un-parsed) expression that's passed to the macro and the whole parsed expression).

ararslan commented on src/formula.jl:
```diff
@@ -54,136 +65,194 @@ function Base.show(io::IO, f::Formula)
-        ## not a :call to s, return condensed version of ex
-        return condense(ex)
+        # recursively sort children
+        sort_terms!.(ex.args)
```
`foreach(sort_terms!, ex.args)` may be more efficient here since it avoids allocations from broadcasting.

kleinschmidt commented:
Ah CategoricalArrays 0.3.8 is going to break this whole package because we rely on unique returning levels order, not order of occurrence.

kleinschmidt commented on REQUIRE:
```diff
@@ -2,3 +2,4 @@ julia 0.6
 DataFrames 0.11.5
 StatsBase 0.20.1
 Compat 0.63
+ArgCheck
```
It's more of a habit than anything for me so I'm fine to do the checks manually.  At this point that wouldn't add too much boilerplate.  Your objection is just that argcheck isn't strictly necessary here, right?

nalimilan replied:
AFAICT you removed all uses of `@argcheck` but still added the dependency.

kleinschmidt replied:
Yes, I caught this right after I tagged a release so I pushed another commit on master to remove it.  Thanks attobot for showing the requires diff.

kleinschmidt commented:
Okay I think #55 will fix the categorical array issue.

nalimilan commented on src/formula.jl:
```diff
@@ -54,136 +65,194 @@ function Base.show(io::IO, f::Formula)
 
-ord(ex::Expr) = (ex.head == :call && ex.args[1] == :&) ? length(ex.args)-1 : 1
-ord(a::Any) = 1
+# ord(ex::Expr) = (ex.head == :call && ex.args[1] == :&) ? length(ex.args)-1 : 1
```
Remove?

nalimilan commented on test/formula.jl:
```diff
@@ -8,10 +8,10 @@
 
     ## totally empty
-    t = Terms(Formula(nothing, 0))
+    t = Terms(@eval @formula $(:($nothing ~ 0)))
```
This is intriguing...

kleinschmidt replied:
well since I've moved parsing into the `@formula` macro and there's not currently syntactic support for one-sided formulae this is what's necessary :)

nalimilan commented on test/formula.jl:
```diff
@@ -101,16 +104,16 @@
 
-    @test dropterm(@formula(foo ~ 1 + bar + baz), :bar) == @formula(foo ~ 1 + baz)
-    @test dropterm(@formula(foo ~ 1 + bar + baz), 1) == @formula(foo ~ 0 + bar + baz)
+    @test_broken dropterm(@formula(foo ~ 1 + bar + baz), :bar) == @formula(foo ~ 1 + baz)
```
I guess you plan to fix this?

kleinschmidt commented:
I've removed the argparse dependency and fixed those tests. I also added a few more parser checks (for `&` with `1` and one-element interactions) that fixes a few tests that were broken before.  Maybe @dmbates wants to weight in before this is merged?  I don't think this should cause any problems with MixedModels but just in case...

dmbates commented:
I do have a problem with one of the tests in `MixedModels` because of the way that I explicitly construct a one-sided formula.  It is at `MixedModels/src/pls.jl:90` and I am trying to build up a model matrix from individual terms.  I use an expression of the form `ModelFrame(Formula(nothing, rhs), fr)` where `rhs` is an expression.

I think it may be best if I rewrite this section of the MixedModels code.  It seems to me that this is probably where https://github.com/dmbates/MixedModels.jl/issues/100 originates.  However, I should have a temporary fix so that this PR in `StatsModels` can go forward.  Is there an explicit constructor for a one-sided formula?

kleinschmidt commented:
You might want to wait to do that re-write because I have bigger changes planned here about how Terms are represented (which may or may not change how best to handle things in MixedModels).

This PR actually fixes that issue, since the parsing rules are applied to the arguments of `|` as well: 

```
julia> @formula( y ~ 1 + x + (a*b | c))
Formula: y ~ 1 + x + ((a + b + a & b) | c)
```

Your issue is due to the fact that the formula holds onto the original expression.  You could probably get away with just constructing a `Formula` with empty `ex` and `ex_orig` fields for now.

dmbates commented:
I had forgotten that the representation of a `Formula` had changed. Would it be possible to add another `Formula` constructor like
```
Formula(lhs::Union{Expr, Symbol, Void}, rhs::Union{Expr, Integer, Symbol}) = Formula(:(), :(), lhs, rhs)
```
to make this case work in the short term?  I could do something in the `MixedModels` package but I think it would be easier to add the method here.

kleinschmidt commented:
I added a `Formula(lhs, rhs)` method that creates a `.ex` field by splicing in the two arguments, and sets `ex_orig` to be `:()`, and gives a deprecation warning suggesting the 4-argument form or the `@eval @formula` solution if parsing is needed.  This ensures backwards compatibility except in cases where you expect the lhs/rhs fields to be _un_-parsed, which...serves them right I guess.

kleinschmidt commented on test/deprecated.jl:
```diff
@@ -1,7 +1,12 @@
     f = @formula y ~ 1 + a*b
 
-    f2 = @test_warn "deprecated" Formula(f.lhs, f.rhs)
+    if VERSION > v"0.7.0-DEV"
```
I mostly want to split 0.6/0.7...I don't know when the `@warn_logs` appeared but I think it's fairly safe to assume that anyone who's on 0.7.something will want the other test.  So somthing like `if VERSION < v"0.6.9999"`?

kleinschmidt replied:
I see that `v"0.7.0"` is higher than any `v"0.7.0-DEV"` so that makes sense to me so this makes sense.

kleinschmidt commented:
I fixed the version check and I think that's the last thing.

matthieugomez commented:
I am trying to update my package for this pull request. 
I have an issue with formula with no left hand side. 
The issue I am encountering is that the following now returns an error:
```julia
using DataFrames
df = DataFrame(x1 = [1])
StatsModels.ModelFrame(StatsModels.Terms(StatsModels.@formula(nothing~x1)), df)
#> ERROR: KeyError: key :nothing not found
```

kleinschmidt commented:
you can do `@eval @formula($nothing ~ x1)`.  If you look at the expression `:(nothing ~ x1)` you'll see that:
```julia
julia> dump( :(nothing ~ x1) )
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Symbol ~
    2: Symbol nothing
    3: Symbol x1
  typ: Any
```
whereas you want
```
julia> dump(:($nothing ~ x1))
Expr
  head: Symbol call
  args: Array{Any}((3,))
    1: Symbol ~
    2: Void nothing
    3: Symbol x1
  typ: Any
```

matthieugomez commented:
Ok. It works now! Thanks a lot for helping me with this.
